src/Liquid/AbstractBlock.php

```diff
@@ -83,15 +89,38 @@ public function parse(array &$tokens)
 					throw new ParseException("Tag $token was not properly terminated (won't match $tagRegexp)");
 				}
 			} elseif ($variableStartRegexp->match($token)) {
+				$this->whitespaceHandler($token);
 				$this->nodelist[] = $this->createVariable($token);
-			} elseif ($token != '') {
+			} else {
+				if (self::$trimWhitespace) {
+					$token = ltrim($token);
+				}
+
+				self::$trimWhitespace = false;
 				$this->nodelist[] = $token;
 			}
 		}
 
 		$this->assertMissingDelimitation();
 	}
 
+	/**
+	 * Handle the whitespace.
+	 *
+	 * @param string $token
+	 */
+	protected function whitespaceHandler(&$token)
+	{
+		if (mb_strlen($token) > 2 && $token[2] === Liquid::get('WHITESPACE_CONTROL')) {
+			$previousToken = end($this->nodelist);
+			if (is_string($previousToken)) {
+				$this->nodelist[key($this->nodelist)] = rtrim($previousToken);
+			}
+		}
+
+		self::$trimWhitespace = (mb_strlen($token) > 2 && substr($token, -3, 1) === Liquid::get('WHITESPACE_CONTROL'));
```

Comment:
I'm having hard time understanding what this function is supposed to do exactly and why. Could you elaborate in a comment?

For example, `mb_strlen` is used here and above, but on the other hand array access operator used which treats all strings as single-byte arrays. It's either `mb_strlen` or `$string[n]` and the common `strlen`, else it makes no sense to me. [A quick demo.](https://3v4l.org/8AVp9)